src/flint/fq_default_mpoly.jl

```diff
@@ -0,0 +1,589 @@
+    return expressify(a.data, x, context = context)
+end
+
+@enable_all_show_via_expressify fq_default_mpoly
+
+
+function show(io::IO, p::FqDefaultMPolyRing)
+    local max_vars = 5 # largest number of variables to print
+    S = symbols(p)
+    n = length(S)
+    print(io, "Multivariate Polynomial Ring in ")
+    if n == 0 || n > max_vars
+        print(io, n)
+        print(io, " variables ")
+    end
+    for i in 1:min(n - 1, max_vars - 1)
+        print(io, string(S[i]), ", ")
+    end
+    if n > max_vars
+        print(io, "..., ")
+    end
+    if n > 0
+        print(io, string(S[n]))
+    end
+    print(io, " over ")
+    print(IOContext(io, :compact => true), base_ring(p))
+end
+
+################################################################################
+#
+#  Getting coefficients
+#
+################################################################################
+
+function coeff(a::fq_default_mpoly, i::Int)
+    return fq_default(base_ring(a), coeff(a.data, i))
+end
+
+function coeff(a::fq_default_mpoly, b::fq_default_mpoly)
+    return fq_default(base_ring(a), coeff(a.data, b.data))
+end
+
+function trailing_coefficient(p::fq_default_mpoly)
+    return fq_default(base_ring(a), trailing_coefficient(p.data))
+end
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+function degree(a::fq_default_mpoly, i::Int)
+    return degree(a.data, i)
+end
+
+function degree_fmpz(a::fq_default_mpoly, i::Int)
+    return degree_fmpz(a.data, i)
+end
+
+function degrees_fit_int(a::fq_default_mpoly)
+    return degrees_fit_int(a.data)
+end
+
+function degrees(a::fq_default_mpoly)
+    return degrees(a.data)
+end
+
+function degrees_fmpz(a::fq_default_mpoly)
+    return degrees_fmpz(a.data)
+end
+
+function total_degree_fits_int(a::fq_default_mpoly)
+    return total_degree_fits_int(a.data)
+end
+
+function total_degree(a::fq_default_mpoly)
+    return total_degree(a.data)
+end
+
+function total_degree_fmpz(a::fq_default_mpoly)
+    return total_degree_fmpz(a.data)
+end
+
+###############################################################################
+#
+#   Multivariable coefficient polynomials
+#
+###############################################################################
+
+function coeff(a::fq_default_mpoly, vars::Vector{Int}, exps::Vector{Int})
+    return fq_default_mpoly(parent(a), coeff(a, vars, exps))
+end
+
+###############################################################################
+#
+#   Basic arithmetic
+#
+###############################################################################
+
+function -(a::fq_default_mpoly)
+    return fq_default_mpoly(parent(a), -a.data)
+end
+
+function +(a::fq_default_mpoly, b::fq_default_mpoly)
+    check_parent(a, b)
+    return fq_default_mpoly(parent(a), a.data + b.data)
```

Comment:
This results in two calls to `jl_apply_generic`. bad?

Reply:
no branches
```
julia> @code_warntype x + y
Variables
  #self#::Core.Const(+)
  a::Nemo.fq_default_mpoly
  b::Nemo.fq_default_mpoly

Body::Nemo.fq_default_mpoly
1 ─      Nemo.check_parent(a, b)
│   %2 = Nemo.parent(a)::Nemo.FqDefaultMPolyRing
│   %3 = Base.getproperty(a, :data)::Union{AbstractAlgebra.Generic.MPoly{fq}, fq_nmod_mpoly, gfp_mpoly}
│   %4 = Base.getproperty(b, :data)::Union{AbstractAlgebra.Generic.MPoly{fq}, fq_nmod_mpoly, gfp_mpoly}
│   %5 = (%3 + %4)::Any
│   %6 = Nemo.fq_default_mpoly(%2, %5)::Nemo.fq_default_mpoly
└──      return %6
```

Reply:
Ah,`@code_warntype` does not show it. I checked and with 3-element union it does not split. With a 2-element union it produces four lovely branches.

Reply:
Typeof(a)==typeof(b), and we don't want 25 branches

Reply:
The other way (without the `typ`), is to do

```
if a.data isa gfp_mpoly && b.data isa gfp_mpoly
...
```

P.S.: Why not use `type` instead of `typ`?